src/domain/app/ssr/commonApolloQueriesService.ts

```diff
@@ -0,0 +1,224 @@
+import { ApolloClient, NormalizedCacheObject } from '@apollo/client';
+import { ArticleType, PageType } from 'react-helsinki-headless-cms';
+import {
+  MenuDocument,
+  MenuQuery,
+  MenuQueryVariables,
+  PagesDocument,
+  PagesQuery,
+  PagesQueryVariables,
+  PostsDocument,
+  PostsQuery,
+  PostsQueryVariables,
+} from 'react-helsinki-headless-cms/apollo';
+
+import { PageUriInfo } from './types';
+import {
+  DEFAULT_HEADER_MENU_NAME,
+  DEFAULT_FOOTER_MENU_NAME,
+  SUPPORTED_LANGUAGES,
+} from '../../../constants';
+
+export const ARTICLES_AMOUNT_LIMIT = 100;
+export const PAGES_AMOUNT_LIMIT = 100;
+
+/**
+ * Service class for executing common Apollo queries related to CMS data.
+ *
+ * NOTE: The queries need to match the real queries that are executed on the client side,
+ * or the cache keys will not match (and therefore will not work as intended).
+ */
+export class CommonApolloQueriesService {
+  /**
+   * The CMS Apollo Client instance used for querying.
+   */
+  private cmsApolloClient: ApolloClient<NormalizedCacheObject>;
+
+  /**
+   * Constructs a new CommonApolloQueriesService instance.
+   *
+   * @param {Object} params - The parameters for the constructor.
+   * @param {ApolloClient<NormalizedCacheObject>} params.cmsApolloClient - The CMS Apollo Client instance.
+   */
+  constructor({
+    cmsApolloClient,
+  }: {
+    cmsApolloClient: ApolloClient<NormalizedCacheObject>;
+  }) {
+    this.cmsApolloClient = cmsApolloClient;
+  }
+
+  /**
+   * Queries and caches the CMS header menu data for one or all supported languages.
+   *
+   * @param {Object} params - The parameters for the query.
+   * @param {SUPPORTED_LANGUAGES} [params.language] - Optional language for the menu.
+   * If omitted, queries for all languages.
+   * @returns {Promise<void>} - A promise that resolves when all queries are complete.
+   */
+  async queryCmsHeaderMenu({
+    language,
+  }: {
+    language?: SUPPORTED_LANGUAGES;
+  }): Promise<void> {
+    const languages: SUPPORTED_LANGUAGES[] = language
+      ? [language]
+      : Object.values(SUPPORTED_LANGUAGES);
+    for (const lang of languages) {
+      const id = DEFAULT_HEADER_MENU_NAME[lang] ?? SUPPORTED_LANGUAGES.FI;
+      if (id) {
+        try {
+          await this.cmsApolloClient.query<MenuQuery, MenuQueryVariables>({
+            query: MenuDocument,
+            variables: {
+              id,
+              menuIdentifiersOnly: true,
+            },
+          });
+        } catch (error) {
+          // eslint-disable-next-line no-console
+          console.error(`Error fetching header menu for ${lang}:`, error);
+          // Or throw the error, or handle it in another way.
+        }
+      } else {
+        // eslint-disable-next-line no-console
+        console.warn('Not fetching a header menu for language', lang);
```

Comment:
How about a more descriptive warning here?
```suggestion
        console.warn(
          `Not fetching header menu for ${lang} because its menu name was empty`
        );
```